actors/builtin/market/market_state.go

```diff
@@ -96,35 +90,37 @@ func (st *State) updatePendingDealState(rt Runtime, state *DealState, deal *Deal
 		st.transferBalance(rt, deal.Client, deal.Provider, totalPayment)
 	}
 
-	if everSlashed {
-		// unlock client collateral and locked storage fee
-		clientCollateral := deal.ClientCollateral
-		paymentRemaining := dealGetPaymentRemaining(deal, state.SlashEpoch)
-		if err := st.unlockBalance(lt, deal.Client, big.Add(clientCollateral, paymentRemaining)); err != nil {
-			rt.Abortf(exitcode.ErrIllegalState, "failed to unlock client balance: %s", err)
-		}
-
-		// slash provider collateral
-		amountSlashed = deal.ProviderCollateral
-		if err := st.slashBalance(et, lt, deal.Provider, amountSlashed); err != nil {
-			rt.Abortf(exitcode.ErrIllegalState, "slashing balance: %s", err)
-		}
-
-		st.deleteDeal(rt, dealID)
-		return amountSlashed, epochUndefined
```

Comment:
I think this introduces a bug. Previously, slashed deals would return here and not enter `processDealExpired` further down. This will now unlock the provider collateral twice.

We need to make sure the expiration processing doesn't do something redundant. I think that either means restoring a flag to the deal state, or deleting the deal state entirely when slashed. Minor preference for the first way.

This might mean that the issue description motivating this change, "Slash deals immediately", was a poor direction. We still need to fix the bug about not slashing deals that are finished, but maybe through a smaller change.

Reply:
@anorth I agree with you. I think keeping the logic in one place also makes it easy to reason about. Have raised #559 .